pkg/controllers/user/logging/clusterlogging.go

```diff
@@ -1,107 +1,118 @@
-		podLister:            cluster.Core.Pods("").Controller().Lister(),
-		services:             cluster.Core.Services(loggingconfig.LoggingNamespace),
-		serviceLister:        cluster.Core.Services("").Controller().Lister(),
-		serviceAccounts:      cluster.Core.ServiceAccounts(loggingconfig.LoggingNamespace),
+		clusterRoleBindings:  cluster.RBAC.ClusterRoleBindings(loggingconfig.LoggingNamespace),
+		clusterLister:        cluster.Management.Management.Clusters("").Controller().Lister(),
 		configmaps:           cluster.Core.ConfigMaps(loggingconfig.LoggingNamespace),
-		namespaces:           cluster.Core.Namespaces(""),
 		daemonsets:           cluster.Apps.DaemonSets(loggingconfig.LoggingNamespace),
 		deployments:          cluster.Apps.Deployments(loggingconfig.LoggingNamespace),
+		deploymentLister:     cluster.Apps.Deployments("").Controller().Lister(),
+		namespaces:           cluster.Core.Namespaces(""),
+		nodeLister:           cluster.Management.Management.Nodes("").Controller().Lister(),
+		podLister:            cluster.Core.Pods("").Controller().Lister(),
+		projectLoggingLister: cluster.Management.Management.ProjectLoggings("").Controller().Lister(),
 		roles:                cluster.RBAC.Roles(loggingconfig.LoggingNamespace),
 		rolebindings:         cluster.RBAC.RoleBindings(loggingconfig.LoggingNamespace),
-		clusterRoleBindings:  cluster.RBAC.ClusterRoleBindings(loggingconfig.LoggingNamespace),
-		clusterLister:        cluster.Management.Management.Clusters("").Controller().Lister(),
-		clusterName:          cluster.ClusterName,
+		services:             cluster.Core.Services(loggingconfig.LoggingNamespace),
+		serviceLister:        cluster.Core.Services("").Controller().Lister(),
+		serviceAccounts:      cluster.Core.ServiceAccounts(loggingconfig.LoggingNamespace),
 	}
-	clusterloggingClient.AddClusterScopedHandler("cluster-logging-controller", cluster.ClusterName, syncer.Sync)
+	clusterloggingClient.AddClusterScopedLifecycle("cluster-logging-controller", cluster.ClusterName, lifecycle)
 }
 
-func (c *ClusterLoggingSyncer) Sync(key string, obj *v3.ClusterLogging) error {
-	//clean up
-	if obj == nil || obj.DeletionTimestamp != nil {
-		return utils.CleanResource(c.namespaces, c.clusterLoggingLister, c.projectLoggingLister)
-	}
+func (c *ClusterLoggingLifecycle) Create(obj *v3.ClusterLogging) (*v3.ClusterLogging, error) {
+	newObj, err := v3.LoggingConditionProvisioned.DoUntilTrue(obj, func() (runtime.Object, error) {
+		return obj, provision(c.namespaces, c.configmaps, c.serviceAccounts, c.clusterRoleBindings, c.daemonsets, c.clusterLister, c.clusterName)
+	})
+	return newObj.(*v3.ClusterLogging), err
+}
 
-	if err := utils.IniteNamespace(c.namespaces); err != nil {
-		return err
-	}
-	if err := utils.InitConfigMap(c.configmaps); err != nil {
-		return err
+func (c *ClusterLoggingLifecycle) Updated(obj *v3.ClusterLogging) (*v3.ClusterLogging, error) {
```

Comment:
@aiwantaozi could you explain the reason why we've switched from sync to a lifecycle?

Reply:
We want to add "Condition Status" for logging, so UI could get the condition and message when deploy failed or pending. Currently, a user can't know the status of the logging components, like embedded logging, if the deployment is pending by lack of resource, user could get the unschedule reason.